### DISCUSSION TITLE
Finding the Best Way to Forward mediamtx Streams to Web Browsers

### DISCUSSION THREAD

**Discussion by Clemens-E** [bluenviron/mediamtx#3400] [Questions]
### Question

Hey there!  Just wanted to say again how awesome this project is – it's constantly blowing me away with its capabilities.

I'm reaching out for some guidance on a potential solution. Here's the scenario:

I have a camera server running mediamtx with multiple cameras connected.
The key goal is to minimize latency while ensuring smooth playback directly in the browser. With that in mind, I tried the mediamtx WebRTC & WHAP connections, and they worked fantastically – virtually no noticeable latency beyond the inherent VPN and internet overhead.

This server is connected to a VM running a REST API via a VPN.
Ideally, I want web browsers to connect to the REST API, discover available cameras and their details, and then be able to view the live streams directly.

By utilizing the REST API as an intermediary, web browsers wouldn't require a VPN connection and would remain completely oblivious to the camera server's IP address and ports. This creates a nice separation of concerns.

Here's a quick diagram to illustrate the flow:
```
             (vpn)       (no vpn)
camera server <=> REST API <=> web browser
```

However, I feel like WebRTC & WHAP might make forwarding data to the browser a bit cumbersome.

Here are some options I've been exploring:

- **Ditch WebRTC:** This would involve using LL-HLS, which would be much easier to implement and manage for proxying the streams. However, the trade-off is a noticeable increase in latency.
- **MediaSoup as SFU:** I considered using something like mediasoup (https://mediasoup.org/) as a Selective Forwarding Unit (SFU). While this seems like a viable approach, the implementation appears quite complex and potentially high-maintenance. If I read correctly, it also does not support WHAP (WHEP) yet
- **Another instance of mediamtx on VM:** This option involves running another instance of mediamtx on the same VM as the REST API. This instance would essentially pull the stream from the mediamtx on the camera server and allow browsers to connect directly via WebRTC. While I haven't tested it yet, it feels like a feasible solution, potentially manageable through API automation.

My questions are:

Is "chaining" mediamtx instances even possible/recommended?
Are there any other approaches I might be overlooking that could achieve my goals?

Any insights or suggestions would be greatly appreciated!

**Comment by aler9** [comment]
Hello,

> Is "chaining" mediamtx instances even possible/recommended?

Absolutely yes, a lot of people do that. You can either forward streams:
https://github.com/bluenviron/mediamtx?tab=readme-ov-file#forward-streams-to-other-servers

Or alternatively use the IP address of an existing MediaMTX instance as RTSP source:
https://github.com/bluenviron/mediamtx?tab=readme-ov-file#rtsp-cameras-and-servers

In this way you can watch streams with WebRTC without exposing the original server to the internet.

> I want web browsers to connect to the REST API, discover available cameras and their details, and then be able to view the live streams directly.

I agree with the idea of using an intermediary backend between MediaMTX and users, in order to filter incoming requests and responses, providing the desired level of security and visibility of the system.

[a comment by an automated account is not shown]
